### ISSUE TITLE
Moving A/B Street crates to this repository

### ISSUE THREAD

**Issue by dabreegster** [a-b-street/osm2streets#41]
@BudgieInWA, I'm thinking about steps to moving `raw_map` and `convert_osm` crates into this repo. Some of the steps would be:

1) Internally split [RawMap](https://github.com/a-b-street/abstreet/blob/master/raw_map/src/lib.rs) into some kind of `StreetNetwork` struct, just with the data/concepts that should live in this repo. Things like buildings, areas, parking lots, the map name, etc would stay in the A/B Street `RawMap`.
2) Internally detangle [convert_osm](https://github.com/a-b-street/abstreet/tree/master/convert_osm/src) into two pieces -- one that should be in this repository, and the A/B Street specific stuff.
3) Split both into separate crates, then move them into this repository.

There would be a dependency loop between the two repos, but I don't think it's a problem, because the crates wouldn't be in a loop. The crates here would only depend on `geom`. The A/B Street side would pull in the osm2streets crates, of course.

The `convert_osm` split will be the more interesting one. There's generic OSM parsing stuff in there. I'll try to keep it as separate as possible -- maybe it's worth a crate of its own, or maybe it's a chance to switch over to something like [osmio](https://github.com/rory/osmio/) and get native PBF reading too.

# Concerns

This will increase my maintenance burden. When this code changes significantly, I regenerate all maps over in A/B Street. To test things temporarily, I'll switch the local dependency in abstreet from this git repo to my local copy. This is a cost that has to be paid eventually, and now's as good a time as any.

But to double check, is this the direction we want to go? We had talked about exploring new schemas and big ideas here and eventually converging, but the current work is building off `RawMap`. I'm eager to try this refactor, but if there's a good chance we won't want it later, no need for churn.

# Questions

1) What should things be called? `RawMap` has always been an awkward choice, just meant to contrast A/B Street's "final" map model. I think we should keep `RawMap` in abstreet (because otherwise there's **lots** of churn renaming files and stuff there) and call the osm2streets bit something else. `StreetNetwork` is my strawman.

2) Most of the data in `RawRoad` and `RawIntersection` looks relevant to move here. The only weird bit is elevation per intersection and percent_incline per road. That's useful in abstreet, but is only populated for a few maps now. OK with keeping the fields in this repo and just not using them much?

# Out of scope for now

Not moving `map_editor` here. It's a useful UI, but has abstreet stuff to create buildings and lots of file IO. An it depends on `widgetry`, which I don't want to tease out yet.

Also not moving the full `importer` tool/pipeline here yet. It does generally useful things like find the  right geofabrik osm.pbf file for a boundary, but it has loads of abstreet-isms. We're already calling `convert_osm` here on `.osm` files -- that's the super simple version of this pipeline.

Also not yet figuring out rendering (#24). There are still big unknowns with cutting over to osm2lanes first (to reap benefits of its separators) and figuring out SVG / widgetry output.

**Comment by dabreegster** [comment]
Oh yeah, and a further step on my mind is how to reconcile abstreet's `map_model::Map` structure with `RawMap`. Rendering code currently takes `map_model` layer stuff, where lanes are split out into individual objects with their own geometry. I'd like to revisit that and possibly keep the street portion of `RawMap` as the final representation. Not sure yet, but getting the clean split here first would be useful to figure out longer term design.

**Comment by BudgieInWA** [comment]
> But to double check, is this the direction we want to go? We had talked about exploring new schemas and big ideas here and eventually converging, but the current work is building off `RawMap`.

Yes, I think this is the right direction. The other option is re-implementing the whole lot bit by bit here, but I think it would be better to avoid that if we can. Having a real consumer of the code is a valuable influence I think. And the wisdom is that quirks in "messy" code bases tend to be hard-earned lessons that any replacements would need to learn too.

With the right part of `RawMap` extracted here, I think it will be feasible to make progress directly on that, instead of building something in parallel. A combination of tests, code reviews, and pinning the abstreet dependency on osm2streets should make this work well.

What about the ability to explore alternative ideas at the same time? Maybe the extracted core lives here under a new name and abstreet keeps `RawMap` as a wrapper of that core. Then I am free to play with another wrapper around that core. Both wrappers will pull the core in slightly different directions. It feels like naming and stuff might be a bit awkward, but its probably inherent awkwardness anyway.

I have been going around in circles trying to figure out what the ideal data representation is for all this stuff, and I think the only fundamental truth is that there isn't one: any representation will make some things easier and other things harder, and I can't find a small enough scope for any of this that lends itself to a "neat" representation. (https://github.com/a-b-street/osm2streets/issues/1#issuecomment-1169667997 is one part of this line of thinking.) My conclusion is that building different "views" of the data -- different abstractions that can be used to interact with the same data -- might be the way to go. If that is the case, then it seems appropriate to do something like: write a wrapper around the extracted core that gives me the `RoadWay`-like "half road" view of things for a transformation algorithm.

> What should things be called?

`StreetNetwork` sounds about right to me. As mentioned, I think `RawMap` should continue to exist as a wrapper of `StreetNetwork` to allow us more freedom in changing `StreetNetwork`.

**Comment by BudgieInWA** [comment]
> Most of the data in `RawRoad` and `RawIntersection` looks relevant to move here.  ... OK with keeping [elevation, incline] in this repo and just not using them much?

I think it is appropriate to bring any and all fields along into this repo, even if they are "cosmetic" as far as the behaviour of this repo is concerned. Carrying those "cosmetic" fields from the OSM data through all of the transformations and associating them with the correct parts of `StreetNetwork` seems like an important capability of this repo.

**Comment by dabreegster** [comment]
Started: https://github.com/a-b-street/abstreet/pull/952

> If that is the case, then it seems appropriate to do something like: write a wrapper around the extracted core that gives me the RoadWay-like "half road" view of things for a transformation algorithm.

+1 to multiple view wrappers over the underlying `StreetNetwork` data. I don't want to hinder your experimentation at all, so let's have the "core" piece live in this repo and continue to be used by "in production" A/B Street stuff. And also have new views / experiments built on top of it be developed here.

> My gut feeling is that an editor should be a non-goal for this repo. Working directly with OSM is my motivation, and I would rather hook osm2streets into JOSM or iD for debugging osm2streets, over building an editor.

Agreed, awesome!

**Comment by dabreegster** [comment]
This is now done! It wasn't so hard after all. I'll file a followup for cleanup
